applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py

```diff
@@ -0,0 +1,185 @@
+
+from .response_function import ResponseFunctionBase
+
+import numpy as np
+
+class ValueAndGradient(object):
+
+    def __init__(self, infeasible_side):
+        self.infeasible_side = infeasible_side
+
+    def GetValue(self, distance, normal):
+        if self.infeasible_side and distance > 0:
+            return self._GetValue(distance, normal)
+        elif not self.infeasible_side and distance < 0:
+            return self._GetValue(distance, normal)
+        return 0.0
+
+    def GetGradient(self, distance, normal):
+        if self.infeasible_side and distance > 0:
+            return self._GetGradient(distance, normal)
+        elif not self.infeasible_side and distance < 0:
+            return self._GetGradient(distance, normal)
+        return [0.0, 0.0, 0.0]
+
+    def _GetValue(self, distance, normal):
+        raise NotImplementedError()
+
+    def _GetGradient(self, distance, normal):
+        raise NotImplementedError()
+
+class ScalarDistancePow(ValueAndGradient):
+
+    exponent = 1
+
+    def _GetValue(self, distance, normal):
+        return pow(distance, self.exponent)
+
+    def _GetGradient(self, distance, normal):
+        factor = self.exponent * pow(distance, self.exponent-1)
+        gradient = [
+            normal[0] * factor,
+            normal[1] * factor,
+            normal[2] * factor
+        ]
+        return gradient
+
+class ComponentDistance(ValueAndGradient):
+    '''Is the same as ScalarDistancePow with exponent 1'''
+
+    def _GetValue(self, distance, normal):
+        return distance
+
+    def _GetGradient(self, distance, normal):
+        gradient = [
+            normal[0],
+            normal[1],
+            normal[2]
+        ]
+        return gradient
+
+class Exponential(object):
+
+    def _GetValue(self, distance, normal):
+        return np.exp(distance)
+
+    def _GetGradient(self, distance, normal):
+        factor = np.exp(distance)
+        gradient = [
+            normal[0] * factor,
+            normal[1] * factor,
+            normal[2] * factor
+        ]
+        return gradient
+
+class PackagingResponseBase(ResponseFunctionBase):
+    """
+
+    Attributes
+    ----------
+    model_part : Model part object of the response function
+    # TODO
+    """
+
+    def __init__(self, identifier, response_settings, model):
+        self.identifier = identifier
+
+        self.response_settings = response_settings
+        self.model = model
+        self.model_part_needs_to_be_imported = False
+
+        model_part_name = response_settings["model_part_name"].GetString()
+        input_type = response_settings["model_import_settings"]["input_type"].GetString()
+        if input_type == "mdpa":
+            self.model_part = self.model.CreateModelPart(model_part_name, 2)
+            domain_size = response_settings["domain_size"].GetInt()
+            if domain_size not in [2, 3]:
+                raise Exception("MassResponseFunction: Invalid 'domain_size': {}".format(domain_size))
```

Comment:
Is this somehow linked to the Mass Response function? In general, I don't quite get the logic of this if section.